Support pre-commit for styling

blue42u opened the pull request:
Currently the canonical way to format and lint any Spack changes is to run `spack style`. This PR adds a slightly more convenient alternative that runs during `git commit` via [`pre-commit`](https://pre-commit.com). Just run `pre-commit install` in your checkouts and never `spack style` again!

The included configuration tries to replicate the effect of `spack style --fix` as closely as possible (`pre-commit` by design is always in `--fix` mode). The command that runs all hooks on all files, `pre-commit run -a`, is added as part of the `prechecks` phase in CI.

This new framework exposes latent linter failures in the codebase. Most (but not all) of these are simply missing type annotations, so this PR also provides simple and unobtrusive fixes to pass pre-commit.

blue42u committed:
Basic pre-commit configuration

blue42u committed:
Add pre-commit to the CI prechecks

becker33 commented on var/spack/repos/builtin/packages/wgrib2/package.py:
```diff
@@ -132,7 +132,13 @@ def inject_flags(self, name, flags):
         return (flags, None, None)
 
-    flag_handler = inject_flags
+    @property
```
`flag_handler` is a method, so I don't think this will work to propagate compiler flags into `wgrib2`.

blue42u replied:
`flag_handler` is defined as a `@property` in `PackageBase` ([ref](https://github.com/spack/spack/blob/develop/lib/spack/spack/package_base.py#L2176-L2184)), so this is needed to bind `self` to the method before returning it as a callable object. Otherwise MyPy errors about `flag_handler` being assigned a callable with a different signature, which I presume is a valid error.

It probably should just be a method, but making it one triggered more MyPy errors in other packages, so :shrug: 

<details>
<summary>Specific errors when `flag_handler` is defined as a method</summary>

```
var/spack/repos/builtin/packages/qmcpack/package.py:216: error: Incompatible types in assignment (expression has type "Callable[[Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]", base class "PackageBase" defined the type as "Callable[[str, Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]")  [assignment]
var/spack/repos/builtin/packages/ipopt/package.py:60: error: Incompatible types in assignment (expression has type "Callable[[Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]", base class "PackageBase" defined the type as "Callable[[str, Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]")  [assignment]
var/spack/repos/builtin/packages/hpctoolkit/package.py:192: error: Incompatible types in assignment (expression has type "Callable[[Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]", base class "PackageBase" defined the type as "Callable[[str, Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]")  [assignment]
var/spack/repos/builtin/packages/hiop/package.py:130: error: Incompatible types in assignment (expression has type "Callable[[Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]", base class "PackageBase" defined the type as "Callable[[str, Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]")  [assignment]
var/spack/repos/builtin/packages/exago/package.py:141: error: Incompatible types in assignment (expression has type "Callable[[Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]", base class "PackageBase" defined the type as "Callable[[str, Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]")  [assignment]
var/spack/repos/builtin/packages/elfutils/package.py:104: error: Incompatible types in assignment (expression has type "Callable[[Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]", base class "PackageBase" defined the type as "Callable[[str, Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]")  [assignment]
var/spack/repos/builtin/packages/upcxx/package.py:104: error: Incompatible types in assignment (expression has type "Callable[[Iterable[str]], Any]", base class "PackageBase" defined the type as "Callable[[str, Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]")  [assignment]
```

</details>

blue42u replied:
New patch lifts `flag_handler` and a few related `@classmethod`'s into proper methods (instead of using Python trickery for nearly the same result).

becker33 commented on lib/spack/spack/builder.py:
```diff
@@ -581,5 +581,9 @@ def __str__(self):
 # Export these names as standalone to be used in packages
-run_after = PhaseCallbacksMeta.run_after
-run_before = PhaseCallbacksMeta.run_before
+def run_after(phase: str, when: Optional[str] = None):
```
Is there a good reason to replace a direct call to the `run_after` method with a wrapper around it?

blue42u replied:
I'm using the wrapper function as a means to add type annotations, without them MyPy errors with:
```
lib/spack/spack/build_systems/_checks.py:128: error: Cannot determine type of "run_after"  [has-type]
```

The alternative is marking it with some long `Callable[...]` type. Would that be preferred?

blue42u replied:
New patch guards these wrapper functions under `if TYPE_CHECKING:`, so they don't cause any slowdown at runtime.

blue42u committed:
Add type annotations to run_after/before

Fixes the following error from MyPy on package files:

    lib/spack/spack/build_systems/_checks.py:128: error: Cannot determine type of "run_after"  [has-type]

blue42u committed:
Remove `from os import symlink` from packages

Fixes the following MyPy errors in packages:

    var/spack/repos/builtin/packages/sombrero/package.py:8: error: Incompatible import of "symlink" (imported name has type "Callable[[Any, Any], Any]", local name has type "Callable[[Union[str, bytes, PathLike[str], PathLike[bytes]], Union[str, bytes, PathLike[str], PathLike[bytes]], bool, DefaultNamedArg(Optional[int], 'dir_fd')], None]")  [assignment]
    var/spack/repos/builtin/packages/smof/package.py:9: error: Incompatible import of "symlink" (imported name has type "Callable[[Any, Any], Any]", local name has type "Callable[[Union[str, bytes, PathLike[str], PathLike[bytes]], Union[str, bytes, PathLike[str], PathLike[bytes]], bool, DefaultNamedArg(Optional[int], 'dir_fd')], None]")  [assignment]
    var/spack/repos/builtin/packages/rmats/package.py:9: error: Incompatible import of "symlink" (imported name has type "Callable[[Any, Any], Any]", local name has type "Callable[[Union[str, bytes, PathLike[str], PathLike[bytes]], Union[str, bytes, PathLike[str], PathLike[bytes]], bool, DefaultNamedArg(Optional[int], 'dir_fd')], None]")  [assignment]
    var/spack/repos/builtin/packages/recon/package.py:8: error: Incompatible import of "symlink" (imported name has type "Callable[[Any, Any], Any]", local name has type "Callable[[Union[str, bytes, PathLike[str], PathLike[bytes]], Union[str, bytes, PathLike[str], PathLike[bytes]], bool, DefaultNamedArg(Optional[int], 'dir_fd')], None]")  [assignment]
    var/spack/repos/builtin/packages/homer/package.py:8: error: Incompatible import of "symlink" (imported name has type "Callable[[Any, Any], Any]", local name has type "Callable[[Union[str, bytes, PathLike[str], PathLike[bytes]], Union[str, bytes, PathLike[str], PathLike[bytes]], bool, DefaultNamedArg(Optional[int], 'dir_fd')], None]")  [assignment]
    var/spack/repos/builtin/packages/fasttree/package.py:8: error: Incompatible import of "symlink" (imported name has type "Callable[[Any, Any], Any]", local name has type "Callable[[Union[str, bytes, PathLike[str], PathLike[bytes]], Union[str, bytes, PathLike[str], PathLike[bytes]], bool, DefaultNamedArg(Optional[int], 'dir_fd')], None]")  [assignment]
    var/spack/repos/builtin/packages/ds/package.py:8: error: Incompatible import of "symlink" (imported name has type "Callable[[Any, Any], Any]", local name has type "Callable[[Union[str, bytes, PathLike[str], PathLike[bytes]], Union[str, bytes, PathLike[str], PathLike[bytes]], bool, DefaultNamedArg(Optional[int], 'dir_fd')], None]")  [assignment]
    var/spack/repos/builtin/packages/circos/package.py:8: error: Incompatible import of "symlink" (imported name has type "Callable[[Any, Any], Any]", local name has type "Callable[[Union[str, bytes, PathLike[str], PathLike[bytes]], Union[str, bytes, PathLike[str], PathLike[bytes]], bool, DefaultNamedArg(Optional[int], 'dir_fd')], None]")  [assignment]
    var/spack/repos/builtin/packages/bridger/package.py:8: error: Incompatible import of "symlink" (imported name has type "Callable[[Any, Any], Any]", local name has type "Callable[[Union[str, bytes, PathLike[str], PathLike[bytes]], Union[str, bytes, PathLike[str], PathLike[bytes]], bool, DefaultNamedArg(Optional[int], 'dir_fd')], None]")  [assignment]

blue42u committed:
Annotate all phases as Tuple[str, ...]

Fixes the following MyPy errors in packages:

    var/spack/repos/builtin/packages/libogg/package.py:39: error: Incompatible types in assignment (expression has type "List[str]", base class "GenericBuilder" defined the type as "Tuple[str]")  [assignment]
    var/spack/repos/builtin/packages/curl/package.py:416: error: Incompatible types in assignment (expression has type "List[str]", base class "NMakeBuilder" defined the type as "Tuple[str, str]")  [assignment]

blue42u committed:
Annotate version table in lammps/package.py

Fixes the following MyPy errors:

    var/spack/repos/builtin/packages/lammps/package.py:489: error: "object" has no attribute "get"  [attr-defined]
    var/spack/repos/builtin/packages/lammps/package.py:491: error: "object" has no attribute "get"  [attr-defined]

blue42u committed:
Lift flag_handler(s) to methods

Fixes the following MyPy error:

    var/spack/repos/builtin/packages/wgrib2/package.py:130: error: Incompatible types in assignment (expression has type "Callable[[Wgrib2, Any, Any], Any]", base class "PackageBase" defined the type as "Callable[[str, Iterable[str]], Tuple[Optional[Iterable[str]], Optional[Iterable[str]], Optional[Iterable[str]]]]")  [assignment]

blue42u commented:
Rebased onto `develop`, and split the MyPy fixes into multiple commits for (hopefully) easier review.

blue42u committed:
Last blackening and flake fix

alecbcs commented:
@becker33
